x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/parser/LogicalPlanBuilder.java

```diff
@@ -531,6 +545,11 @@ private Expression pipeIntArgument(Source source, String pipeName, List<BooleanE
             );
         }
 
+        if (expression.dataType() != INTEGER && expression.dataType() != SHORT && expression.dataType() != BYTE) {
+            // 2147483650 - 4 will yield an integer (Integer.MAX_VALUE - 1) but the expression itself (SUB) will be of type LONG
+            // and this type will be used when being folded later on
+            return new Literal(expression.source(), value.intValue(), INTEGER);
+        }
         return expression;
```

Comment:
question: could we replace this by
```suggestion
        return new Literal(expression.source(), value.intValue(), expression.dataType());
```
to leverage the fact that we already folded the expression?

Reply:
`expression.dataType()` can be something different than INTEGER. See https://github.com/elastic/elasticsearch/pull/101859/files#diff-123105fcd0030e62f4100ab1424e1e978bcffa93eb64169867557eb3136c0fe1R549

Reply:
I've changed it to always return an INTEGER Literal, since we already folded the value.
But this is not entirely correct, because the folding itself should be dealt with in the Optimizer.ConstantFolding rule. But there are situations (`head number("123")`) where the function itself (`number`) hasn't been resolved yet in order to be folded.